Use sqlite

Profpatsch opened the pull request:
<!--
Explain the approach you took to resolving the issue and provide necessary context.
There is no need to go into a lot of detail here: instead, try to make each commit self-explanatory and write good commit messages.
-->

- [ ] Amended the changelog in `release.nix` (see `release.nix` for instructions)

Profpatsch committed:
feat: set up a simple rusqlite example

Profpatsch committed:
feat(sqlite): start importing the gcroots data

This is super crappy at the moment, but we want to ideally save most
of the data in the database instead of indirectly in the filesystem.
This gives us better leverage for storing other fun things, like which
files we last watched for a project, or how long changes to the files
took.

nyarly commented:
I'm not following the motivation here.

Based on the commit messages, I can see some ideas about potential features that would come out of this, but I think the folks who are or might use Lorri are looking for two things: GC root setups and flakes and I'd like to prioritize getting those done.

Profpatsch commented:
> GC root setups and flakes and I'd like to prioritize getting those done.

Yeah, I feel like flakes are kind of important; but I personally don’t use them yet, so I don’t have a huge motivation to work on them. I’ll of course review any changes to them.



> I'm not following the motivation here.

The motivation is that I want lorri to keep around info between runs, more than the current GC roots. And a small database file is the easiest setup for that.

In particular, adding timings to previous evaluation and build times of files, so that lorri can estimate how long a change is gonna take to build.

Profpatsch committed:
refact(sqlite): make struct, add in_transaction, save to file

Profpatsch committed:
feat(project): start saving nix_file in database

In addition to updating the backlink to nix_file, we also update the gc_roots
table in our database to contain the same info.

The next step is to read everything from the database instead of from the
symlinks.

Currently, the gc_roots table is deleted at every start (to make changes to the
schema easier), after everything works through the database, we have to set up a
one-time migration (and write a test for it …) and clean up the symlinks.

Profpatsch committed:
refact(project): add connection to struct

Luckily the `Connection` type of rusqlite is `Clone` now (because it
only has to clone the mpsc sender instead of the actual sqlite
connection object), so we can just make everything `Clone` yay.

Profpatsch commented:
Another thing I noticed when rebasing this yesterday, for flakes we have to store the `Installable` string somewhere so we can know which part of the flake we need to use.

Profpatsch committed:
fix(daemon): replace forgotten todo! macro

Profpatsch committed:
refact(*): Use tokio current thread runtime everywhere

(diff best viewed without whitespace changes)

lorri does not really need the multi-threaded runtime.
Instead, we can move everything to use `spawn_local`, meaning we don’t
need a `Send` constraint on our futures, because no value will be
transferred between threads, major profit.

Annoyingly, the `tokio::test` macro does not support putting things on
a `LocalSet` or `LocalRuntime` for us, so we have to wrap everything
in our setup manually. But that’s fine.

I rewrite the `.retain` logic to use optional channels that get
set to None and cleaned of those afterwards, so we can .await on the
channel send.

Profpatsch committed:
refact(project): move Installable from lib to project

Profpatsch committed:
doc(project): adds some documentation to the flake output & rename

Profpatsch committed:
refact(builder): inline extract_paths

Profpatsch committed:
refact(project): specialize the extra_paths to the one we need

We only need it for the single flake profile path, so let’s not
overgeneralize this before we have any idea whether we need more than
a single extra one.

Also add more docs and rename a few things.

Profpatsch committed:
refact(project): move lorri_gc_root out of create_root

Profpatsch committed:
refact(project): Simplify the gc root path generation

There’s little reason to do `base_name` bs of static filenames that
were generated somewhere else, we know there’s exactly one name for
flake profiles.

This renames the `profile` symlink to `flake_profile`, which should be
fine because we don’t rely on the name anywhere.

fwiw the `remove_project` gc function still works, because it deletes
the whole project gc dir, which contains both the `shell_gc_root` and
`profile` link.

Profpatsch committed:
refact(project): project.hash is never read

We generate the project from the filename now, so we never need the
hash apart from generating the filename after hashing the shell.nix or
flake.nix.

Profpatsch committed:
refact(project): don’t split path unnecessarily

We don’t have to split the hash from the file path, we can just pass
the path into `new_internal_from_existing_gc_dir`, which only concats
it in the end.

Profpatsch committed:
refact(project): inline new_internal_from_existing_gc_dir

Profpatsch committed:
chore(main): remove is_file_in_current_directory

This was only used in a single test in main, and in general this whole
test is a head-scratcher. What’s the thing being tested here? Not
sure.

At least let’s get rid of the weird dependency on AbsDirPathBuf.

Profpatsch committed:
fix(ops): fix json path output for gc_dir

Profpatsch committed:
refact(lib): replace AbsDirPathBuf with AbsPathBuf

We don’t really need AbsDirPathBuf here. I think we can just remove it
completely. Should have the same semantics and be clear.

Profpatsch committed:
refact(project): inline last_built_timestamp

Profpatsch committed:
refact(project): inline list_roots_impl

Inline it twice, because we need to keep the from-file reader for the
migration part and translate the rest to read from sqlite.

Profpatsch commented:
![image](https://github.com/user-attachments/assets/e3515232-71a1-4299-a5d4-1affb840d298)

Profpatsch committed:
feat(src/project): list GC roots from sqlite database

We need to keep the original function around, because that’s how we
migrate to the sqlite database in the first place. Maybe we can
simplify it though?

We definitely need a test that checks the migration path.

Profpatsch committed:
feat(src/project): add test for creation of indirect gc roots

This tests the creation of gc roots pretty directly.

Only problem is that `nix-store --delete` needs to list
`/nix/store/.links` every time, this is 1.5 million files on my
system, which takes about 5 seconds, and the test has to run it twice
ugh. Not sure what to do about that.

Profpatsch committed:
chore(src/project): move struct down

Profpatsch committed:
feat(sqlite): test the migration to sqlite

Adds a test that assures the migration code works according to the
existing gc dirs.

Now, this does construct a separate gc dir, but since we are gonna
move the `project` module away from doing things in the file system as
much, we will have to manually construct a pre-migration dir anyway
for the migration test.

This should test all stored bits from the gc dirs, and compare the
migration of both flakes and shell.nix files.

Profpatsch committed:
fix(project): remove the nix_file backlink & gc from db

We don’t need the nix file gc backlink anymore, since we list the
original path in the database.

Also fix the horrible setup for the gc integration test, use the test
paths thingy instead of trying to reset all XDG variables like some
cavemen.

We delete the nix file during migration in any setups where it’s still
around.

Profpatsch committed:
feat(sqlite): finish migrating to sqlite

This finishes the initial migration to sqlite.

1) We add a check whether we are already migrated.
2) If we successully manage to migrate, we delete the nix file
backlinks to clean up our cache directories.

The migration test now checks that the backlinks are sucessfully
removed as well.

The database file is never cleaned, and renamed to `lorri.sqlite`.

From this change on, we’ll have to think about and test migration
paths between database versions, somehow …

Profpatsch commented:
Okay, this should now be ready for testing. I added a bunch of tests, but ideally someone else can verify that this works.

NB: backup your `~/.cache/lorri`, because I remove the backlinks to the nix files, meaning you can’t roll back if the migration goes through.

Profpatsch commented:
I’m gonna merge now, if there’s any issues please open an issue (heh).

The next release should not be done right away, but when there’s actual new features or important bug fixes. Maybe a point-release from before this PR could be made.

nyarly commented:
I'll move my testing branch to this and check it out.
